plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergBucketFunction.java

```diff
@@ -79,6 +89,35 @@ public IcebergBucketFunction(
                 .collect(toImmutableList());
     }
 
+    private static void addFieldInfo(
+            Map<String, FieldInfo> nameToFieldInfo,
+            int sourceChannel,
+            String fieldName,
+            List<ColumnIdentity> children,
+            Type type,
+            LinkedList<Integer> path)
+    {
+        if (type instanceof RowType rowType) {
+            List<RowType.Field> fields = rowType.getFields();
+            checkArgument(children.size() == fields.size(), format("children size (%s) == fields size (%s) is not equal", children.size(), fields.size()));
+            for (int i = 0; i < fields.size(); i++) {
+                ColumnIdentity child = children.get(i);
+                String qualifiedFieldName = fieldName + "." + child.getName();
+                path.add(i);
+                if (fields.get(i).getType() instanceof RowType) {
+                    addFieldInfo(nameToFieldInfo, sourceChannel, qualifiedFieldName, child.getChildren(), fields.get(i).getType(), path);
+                }
+                else {
+                    nameToFieldInfo.put(qualifiedFieldName, new FieldInfo(sourceChannel, qualifiedFieldName, fields.get(i).getType(), ImmutableList.copyOf(path)));
+                }
+                path.removeLast();
+            }
+        }
+        else {
+            nameToFieldInfo.put(fieldName, new FieldInfo(sourceChannel, fieldName, type, ImmutableList.copyOf(path)));
+        }
+    }
+
     @Override
     public int getBucket(Page page, int position)
```

Comment:
We've discussed earlier that https://github.com/trinodb/trino/pull/12734 could help in producing a more efficient batched implementation here.
Can we do some simple benchmarks like comparing CPU and latency of partitioning by a primitive column vs the same column stored as a field in a ROW type with current PR ?

Reply:
ping @krvikash @findinpath @ebyhr 
Can we do a simple benchmark comparing the impact of partitioning by a primitive column vs the same column stored as a field in a ROW type with current PR ? We should have a rough idea about how it compares so that we know whether this case needs to be improved and whether perf sensitive users should avoid partitioning on a field in a ROW type.

Reply:
@raunaqmorarka  I ran the following statement 5 times with 2 `partitioning` table properties. The performance seems the same. 
``` sql
 CREATE TABLE iceberg.tpch.tpch_sf100_orders
 WITH (
    format = 'PARQUET',
    partitioning = ARRAY['orderdate'] /* or ARRAY['"nested.dt"'] */
 )
 AS SELECT *, CAST(ROW(orderdate) AS ROW(dt DATE)) AS nested FROM tpch.sf10.orders;
```

partitioning = ARRAY['orderdate']`
```
Splits: 95 total, 95 done (100.00%)
CPU Time: 85.8s total,  175K rows/s,     0B/s, 49% active
Per Node: 1.7 parallelism,  297K rows/s,     0B/s
Parallelism: 5.1
Peak Memory: 2.1GB
16.82 [15M rows, 0B] [892K rows/s, 0B/s]

Splits: 95 total, 95 done (100.00%)
CPU Time: 86.7s total,  173K rows/s,     0B/s, 48% active
Per Node: 1.8 parallelism,  311K rows/s,     0B/s
Parallelism: 5.4
Peak Memory: 1.96GB
16.10 [15M rows, 0B] [932K rows/s, 0B/s]

Splits: 95 total, 95 done (100.00%)
CPU Time: 85.6s total,  175K rows/s,     0B/s, 49% active
Per Node: 1.8 parallelism,  311K rows/s,     0B/s
Parallelism: 5.3
Peak Memory: 2.05GB
16.08 [15M rows, 0B] [933K rows/s, 0B/s]

CPU Time: 86.3s total,  174K rows/s,     0B/s, 50% active
Per Node: 1.9 parallelism,  325K rows/s,     0B/s
Parallelism: 5.6
Peak Memory: 2.04GB
15.39 [15M rows, 0B] [974K rows/s, 0B/s]

Splits: 95 total, 95 done (100.00%)
CPU Time: 86.7s total,  173K rows/s,     0B/s, 51% active
Per Node: 1.8 parallelism,  311K rows/s,     0B/s
Parallelism: 5.4
Peak Memory: 2.01GB
16.10 [15M rows, 0B] [932K rows/s, 0B/s]
```

partitioning = ARRAY['"nested.dt"']
```
Splits: 95 total, 95 done (100.00%)
CPU Time: 85.6s total,  175K rows/s,     0B/s, 50% active
Per Node: 1.9 parallelism,  327K rows/s,     0B/s
Parallelism: 5.6
Peak Memory: 2.04GB
15.29 [15M rows, 0B] [981K rows/s, 0B/s]

Splits: 95 total, 95 done (100.00%)
CPU Time: 84.9s total,  177K rows/s,     0B/s, 50% active
Per Node: 1.8 parallelism,  322K rows/s,     0B/s
Parallelism: 5.5
Peak Memory: 2.05GB
15.52 [15M rows, 0B] [967K rows/s, 0B/s]

Splits: 95 total, 95 done (100.00%)
CPU Time: 87.4s total,  172K rows/s,     0B/s, 51% active
Per Node: 1.9 parallelism,  323K rows/s,     0B/s
Parallelism: 5.6
Peak Memory: 2.06GB
15.48 [15M rows, 0B] [969K rows/s, 0B/s]

Splits: 95 total, 95 done (100.00%)
CPU Time: 87.2s total,  172K rows/s,     0B/s, 51% active
Per Node: 1.9 parallelism,  331K rows/s,     0B/s
Parallelism: 5.8
Peak Memory: 2.14GB
15.08 [15M rows, 0B] [994K rows/s, 0B/s]

Splits: 95 total, 95 done (100.00%)
CPU Time: 88.3s total,  170K rows/s,     0B/s, 51% active
Per Node: 1.9 parallelism,  325K rows/s,     0B/s
Parallelism: 5.7
Peak Memory: 2.01GB
15.41 [15M rows, 0B] [974K rows/s, 0B/s]
```